cedar-policy-generators/src/schema.rs

```diff
@@ -640,74 +640,84 @@ impl Schema {
         };
         let mut principal_types = HashSet::new();
         let mut resource_types = HashSet::new();
+        // optionally return a list of entity types and add them to `tys` in the same time
+        let pick_entity_types = |tys: &mut HashSet<Name>, u: &mut Unstructured<'_>| {
+            Result::Ok(if u.ratio::<u8>(1, 4)? {
+                // The action applies to an unspecified
+                // resource and no other entity types.
+                None
+            } else {
+                // for each entity type, flip a coin to see
+                // whether to include it as a possible
+                // principal type for this action
+                Some(
+                    entity_types
+                        .iter()
+                        .filter_map(|(name, _)| match u.ratio::<u8>(1, 2) {
```

Comment:
[nit] Maybe worth using named constants for these

Reply:
I think I'll keep this way because we've got ratios like this in several places.